### PR TITLE
Context manager example and tests

### PR DESCRIPTION
**Pull request by mosquito**
Add context manager behaviour for channel.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -120,6 +120,43 @@ which is functionally equivalent to
...
+               # Create ten asyncronous tasks
+               workers = [
+                   event_loop.create_task(worker(channel))
+                   for _ in range(10)
+               ]
+
+               for index in range(100):
+                   await channel.put(str(index))
+
+           # Now we are waiting then for all workers exits
```
Comment (tzoiker) [inline README.rst]: `Now we are waiting for all workers to exit`

File: README.rst [hunk 2]
```diff
@@ -120,6 +120,43 @@ which is functionally equivalent to
...
+           print("Channel closed")
+
+
+       async def main():
+           # Hint the type here
+           channel: Channel[str]
+           async with Channel(10) as channel:
+               event_loop = asyncio.get_event_loop()
+
+               # Create ten asyncronous tasks
```
Comment (tzoiker) [inline README.rst]: `asynchronous`

**Comment by mosquito** [comment]
Perhaps fixes #13, probably 😄

**Review by dizballanze — DISMISSED** [review]

**Comment by mosquito** [comment]
@tbug I found a very simple way to make code with several asynchronous tasks easier and cleaner, please take a look.

(By the way, I think it's time to release.)

**Comment by tudborg** [comment]
Thanks @mosquito 👍 

I've deliberately not added this, because I wasn't sure what semantics would be useful for an async context manager.
I generally tend to not add abstraction, if all it saves me is a line of code.

Does this add enough value compared to just calling close and join manually?
Could this be the cause of hard-to-find bugs?

My thought is that if you use the async context manager, then `__aexit__` will close and join,
but if you have an exception early inside your context, before you spawn any workers,
this will hang forever, waiting for the channel to drain, which will never happen.


Could you justify in what scenario a context manager would be better than:

```python
channel = Channel(10)
workers = [loop.create_task(worker(channel)) for _ in range(10)]
for item in range(1000):
    await channel.put(item)
channel.close()
channel.join()
```

**Comment by mosquito** [comment]
The reason is simple, we are all human and we all make mistakes. For example, you forgot `await` in `channel.join()` in your example.

In terms of one line, you're right, but the full equivalent of `__aexit__` is:

```python
channel = Channel(10)
workers = [loop.create_task(worker(channel)) for _ in range(10)]

try:
    for item in range(1000):
        await channel.put(item)
finally:
    channel.close()
    await channel.join()
```

This example is a little fanciful because there is no place where something can go wrong, but in real life if something goes wrong in a task which adding to a channel, one should almost always close the channel and crash.

IMHO: the `async with` semantic is a really simple to understand. Exactly similar reason of existance context manager for a file object, and in this case the only one line with `.close()` should be added.

**Review by Alviner — APPROVED** [review]

**Comment by tudborg** [comment]
> Exactly similar reason of existance context manager for a file object

I disagree that these are comparable, and that might be why I don't understand the value.
With a file object or similar, the responsibility of finalizing the context is held by the context manager.
If any exception happens in the context body, you totally want to close the file.
But closing the file can be done from within the context manager exit. It doesn't have to cooperate with any other code to close the resource handle to the file.

This is different than with a Channel, where exiting a context could easily hang forever if there is no one else available to drain the channel.
This is why I think this rendition of `__aenter__`/`__axit__` is unsafe.

Continuing the example:

```python
channel = Channel(10)
workers = [loop.create_task(worker(channel)) for _ in range(10)]

try:
    raise RuntimeError("imagine some code would raise here")
    for item in range(1000):
        await channel.put(item)
finally:
    channel.close()
    await channel.join()
```

In this tiny example it's pretty easy to avoid calling code that raises here.
But you cannot know that this is important if you hide this behaviour (specifically, the call to `join` inside the context exit)

And it's not just a problem only with other peoples code. It would be easy to make the following mistake with just aiochannel code:

Assuming a context manager with a join in the exit, any call that raises an exception could potentially cause an unexpected hang.

the following examples would hang forever:

```python
with Channel(1) as channel:
    channel.get_nowait()
```

```python
with Channel(1) as channel:
    channel.close()
    channel.get()
```

```python
with Channel(1) as channel:
    channel.put_nowait()
    channel.put_nowait()
```

etc.

As to your comment on the forgotten `await`, adding a context manager around the channel would neither help nor worsen that situation. Types, tests, and running with asyncio warnings for un-`await`ed returns would. 

---

This ofc. doesn't mean that there shouldn't be a context manager.
My point is that if we are adding a context manager that has pitfalls,
it also better solve some real world issues.
The justifications you gave for this addition was:

> The reason is simple, we are all human and we all make mistakes.

and

> the async with semantic is a really simple to understand. Exactly similar reason of existance context manager for a file object, and in this case the only one line with .close() should be added.

Neither of which gives any real reason for the addition.

I very much agree with your comment that

> but in real life if something goes wrong in a task which adding to a channel, one should almost always close the channel and crash.

This is very true, you want to crash.
But if you have a `join` in your `__aexit__`, and an exception happens before you can instantiate any draining tasks, then your `__aexit__` would hang forever.

To me, this makes this currently proposed context manager usafe to use.


To reiterate the question I posed:

Does this add enough value compared to just calling close and join manually,
especially considering the added ability to crate hard-to-find bugs?
And more specifically, is this change solving a concrete problem you have?

There might be a use-case I'm missing here.
You'll have to explain why this addition is helpful to you.

**Pull request closed without merging**